README.md
[![Platforms](https://img.shields.io/endpoint?url=https%3A%2F%2Fswiftpackageindex.com%2Fapi%2Fpackages%2Fbircni%2FBetterTabBar%2Fbadge%3Ftype%3Dplatforms)](https://swiftpackageindex.com/bircni/BetterTabBar)
[![Swift Versions](https://img.shields.io/endpoint?url=https%3A%2F%2Fswiftpackageindex.com%2Fapi%2Fpackages%2Fbircni%2FBetterTabBar%2Fbadge%3Ftype%3Dswift-versions)](https://swiftpackageindex.com/bircni/BetterTabBar)
[![Build](https://github.com/bircni/BetterTabBar/actions/workflows/swift.yml/badge.svg?branch=main)](https://github.com/bircni/BetterTabBar/actions/workflows/swift.yml)

# BetterTabBar

Enhance your App with a better TabBar!

<p align="center">
<img src="./docs/assets/example.gif" alt="Example" style="zoom:70%;"
</p>

## Installation

### Swift Package Manager

Add the following to your `Package.swift` file:

```swift
let package = Package(
    ...
    dependencies: [
        ...
        .package(name: "BetterTabBar", url: "https://github.com/bircni/BetterTabBar.git", from: "0.1.0"),
    ],
    targets: [
        .target(
            name: "MyApp",
            dependencies: ["BetterTabBar"] // Add as a dependency
        )
    ]
)
```

### Add package dependency in Xcode

Use this [guide](https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app) to add `BetterTabBar` to your project. Use <https://github.com/bircni/BetterTabBar> for the URL when Xcode asks.

## Usage

Use the Example in this repository as a starting point.

**Create a main View and add the `BetterTabBar`**

```Swift
import SwiftUI
import BetterTabBar

struct ContentView: View {
    @AppStorage("tabId") var tabId: Int = 0

    var body: some View {
        ZStack (alignment: .bottom){
            switch tabId {
            case 0:
                VStack {
                    Spacer()
                    Text("View 1")
                    Spacer()
                }
            case 1:
                VStack {
                    Spacer()
                    Text("View 2")
                    Spacer()
                }
            case 2:
                VStack {
                    Spacer()
                    Text("View 3")
                    Spacer()
                }
            case 3:
                VStack {
                    Spacer()
                    Text("View 1")
                    Spacer()
                }
            default:
                Text("View 1")
            }
            
            BetterTabBar(tabId: $tabId, strokeColor: .blue, strokeLineWidth: 2, spacing: 25, cornerRadius: 30, height: 60, tapFeedback: true, buttons: [
                TabButton(selectedImage: Image(systemName: "bus"), unselectedImage: Image(systemName: "bus.fill"), lightColor: .green, darkColor: .blue),
                TabButton(selectedImage: Image(systemName: "car"), unselectedImage: Image(systemName: "car.fill"), lightColor: .green, darkColor: .blue),
                TabButton(selectedImage: Image(systemName: "tram"), unselectedImage: Image(systemName: "tram.fill"), lightColor: .green, darkColor: .blue)
            ])
        }
    }
}
```

## Apps using BetterTabBar

- [ParkenUlm](https://apps.apple.com/de/app/id6447651336)
- [FuelLogger](https://apps.apple.com/app/id6449910025)
- [DUO](https://testflight.apple.com/join/0Sm4IXio)

## Contributing

Contributions are very welcome 🙌

To submit your changes, please open a Pull Request with a clear description of your changes.

## Contributors

<p align="center">
<a href="https://github.com/bircni/BetterTabBar/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=bircni/BetterTabBar"/>
</a>
</p>


CHANGELOG.md
# 0.0.1

* first pre-release


Tests/BetterTabBarTests/BetterTabBarTests.swift
import XCTest
@testable import BetterTabBar

final class BetterTabBarTests: XCTestCase {
    func testExample() throws {
        // XCTest Documentation
        // https://developer.apple.com/documentation/xctest

        // Defining Test Cases and Test Methods
        // https://developer.apple.com/documentation/xctest/defining_test_cases_and_test_methods
    }
}


Package.swift
// swift-tools-version: 5.7
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "BetterTabBar",
    platforms: [
        .iOS(.v15),
        .macOS(.v12),
        .macCatalyst(.v15)
    ],
    products: [
        // Products define the executables and libraries a package produces, making them visible to other packages.
        .library(
            name: "BetterTabBar",
            targets: ["BetterTabBar"]), //, "TabButton"
    ],
    targets: [
        // Targets are the basic building blocks of a package, defining a module or a test suite.
        // Targets can depend on other targets in this package and products from dependencies.
        .target(
            name: "BetterTabBar"),
    ]
)


Example/Example/ContentView.swift
//
//  ContentView.swift
//  Example
//
//  Created by bircni
//

import SwiftUI
import BetterTabBar

struct ContentView: View {
    @AppStorage("tabId") var tabId: Int = 0

    var body: some View {
        ZStack (alignment: .bottom){
            switch tabId {
            case 0:
                VStack {
                    Spacer()
                    Text("View 1")
                    Spacer()
                }
            case 1:
                VStack {
                    Spacer()
                    Text("View 2")
                    Spacer()
                }
            case 2:
                VStack {
                    Spacer()
                    Text("View 3")
                    Spacer()
                }
            case 3:
                VStack {
                    Spacer()
                    Text("View 1")
                    Spacer()
                }
            default:
                Text("View 1")
            }
            
            BetterTabBar(tabId: $tabId, strokeColor: .blue, strokeLineWidth: 2, spacing: 25, cornerRadius: 30, height: 60, tapFeedback: true, buttons: [
                TabButton(selectedImage: Image(systemName: "bus"), unselectedImage: Image(systemName: "bus.fill"), lightColor: .green, darkColor: .blue),
                TabButton(selectedImage: Image(systemName: "car"), unselectedImage: Image(systemName: "car.fill"), lightColor: .green, darkColor: .blue),
                TabButton(selectedImage: Image(systemName: "tram"), unselectedImage: Image(systemName: "tram.fill"), lightColor: .green, darkColor: .blue)
            ])
        }
    }
}

#Preview {
    ContentView()
}


Sources/BetterTabBar/BetterTabBar.swift
import Foundation
import SwiftUI

public struct BetterTabBar: View {
    @Environment(\.colorScheme) var colorScheme
    @Binding var tabId: Int
    var strokeColor: Color
    var strokeLineWidth: CGFloat
    var spacing: CGFloat
    var maxWidth: CGFloat?
    var cornerRadius: CGFloat
    var height: CGFloat
    var tapFeedback: Bool
    var buttons: [TabButton]
    
    public init(tabId: Binding<Int>, strokeColor: Color, strokeLineWidth: CGFloat, spacing: CGFloat, maxWidth: CGFloat? = nil, cornerRadius: CGFloat, height: CGFloat, tapFeedback: Bool, buttons: [TabButton]) {
        self._tabId = tabId
        self.strokeColor = strokeColor
        self.strokeLineWidth = strokeLineWidth
        self.spacing = spacing
        self.maxWidth = maxWidth
        self.cornerRadius = cornerRadius
        self.height = height
        self.tapFeedback = tapFeedback
        self.buttons = buttons
    }
    
    public var body: some View{
        VStack (spacing: 15) {
            RoundedRectangle(cornerRadius: cornerRadius)
                .fill(.regularMaterial)
                .frame(maxWidth: maxWidth?.isNaN ?? true ? CGFloat(60 * buttons.count) : maxWidth)
                .overlay(
                    HStack (alignment: .center, spacing: spacing) {
                        ForEach(Array(buttons.enumerated()), id: \.element.id) { index, button in
                            let image = tabId == index ? button.selectedImage : button.unselectedImage
                            Button {
                                if tapFeedback {
#if os(iOS)
                                    UIImpactFeedbackGenerator(style: .medium).impactOccurred()
#endif
                                }
                                withAnimation {
                                    tabId = index
                                }
                            } label: {
                                image
                                    .font(.system(size: 25, weight: tabId == index ? .semibold : .regular))
                                    .foregroundColor(colorScheme == .light ? button.lightColor : button.darkColor)
                                    .background(Color.clear)
                                    .opacity(tabId == index ? 1 : 0.5)
                            }
                        }
                    }
                )
                .overlay(
                    RoundedRectangle(cornerRadius: cornerRadius)
                        .stroke(strokeColor.opacity(0.6), lineWidth: strokeLineWidth)
                )
                .frame(height: height)
                .animation(.spring())
        }
    }
}

public struct TabButton {
    public let id: UUID = UUID()
    public var selectedImage: Image
    public var unselectedImage: Image
    public var lightColor: Color
    public var darkColor: Color
    
    public init(selectedImage: Image, unselectedImage: Image, lightColor: Color, darkColor: Color) {
        self.selectedImage = selectedImage
        self.unselectedImage = unselectedImage
        self.lightColor = lightColor
        self.darkColor = darkColor
    }
}


Examples/Example.swift
//
//  SwiftUIView.swift
//  
//
//  Created by Nicolas Bircks on 04.11.23.
//

import SwiftUI
import BetterTabBar

@available(iOS 15.0, *)
struct Example: View {
    @AppStorage("tabId") var tabId: Int = 0
    var body: some View {
        ZStack (alignment: .bottom){
            switch tabId {
            case 0:
                VStack {
                    Spacer()
                    Text("View 1")
                    Spacer()
                }
            case 1:
                VStack {
                    Spacer()
                    Text("View 2")
                    Spacer()
                }
            case 2:
                VStack {
                    Spacer()
                    Text("View 3")
                    Spacer()
                }
            case 3:
                VStack {
                    Spacer()
                    Text("View 1")
                    Spacer()
                }
            default:
                Text("View 1")
            }
            CustomTabBar1(tabId: $tabId, strokeColor: .blue, strokeLineWidth: 2, spacing: 25, cornerRadius: 30, height: 60, buttons: [
                .init(selectedImage: Image(systemName: "bus"), unselectedImage: Image(systemName: "bus.fill"), lightColor: .green, darkColor: .blue),
                .init(selectedImage: Image(systemName: "car"), unselectedImage: Image(systemName: "car.fill"), lightColor: .green, darkColor: .blue),
                .init(selectedImage: Image(systemName: "tram"), unselectedImage: Image(systemName: "tram.fill"), lightColor: .green, darkColor: .blue)
            ])
        }
    }
}

@available(iOS 15.0, *)
#Preview {
    Example()
}
